alsa-tools/sb16_csp/cspctl.1
.TH cspctl 1 "April 27, 2000"
.SH NAME
cspctl \- Sound Blaster 16 ASP/CSP control program

.SH SYNOPSIS
.B cspctl
command [\-options] <filename>

.SH DESCRIPTION
.B cspctl
is a simple CSP microcode loader for Creative Sound Blaster 16ASP and some Sound Blaster AWE32 sound cards with Creative Signal Processor (CSP) chip (CT1748A) installed.
.B cspctl
can also be used as a post-install function to
.I snd\-sb16\-csp
module to load default codec at module installation. If CSP chip is successfully detected and initialized, it will be installed as a hardware dependent device
.I hwC0D2
into /dev/snd directory. Currently, following codecs can be loaded to CSP device:
.IP "" 4
.Sp
.nf
.ta \w'\fBwfm0000a.csp\fP  'u
\&\fBwfm0001a.csp\fP	QSound decoder
\&\fBwfm0006a.csp\fP	A\-law codec
\&\fBwfm0007a.csp\fP	u\-law codec
\&\fBwfm0011a.csp\fP	IMA ADPCM codec (distorted output for IMA test files)
\&\fBwfm0200a.csp\fP	Creative ADPCM codec (sounds like Intel/DVI IMA ADPCM compatible)
\&\fBwfm0202a.csp\fP	Fast Speech 8 codec
\&\fBwfm0203a.csp\fP	Fast Speech 10 codec
.Sp
.fi
.PP
These codecs are not yet supported by ALSA:
.IP "" 4
.ta \w'\fBwfm0000a.csp\fP  'u
\&\fBwfm0201a.csp\fP	Text2Speech decoder
.Sp
.fi
.PP
Qsound microcode file has
.B four functions,
each function is a decoder for different sample rate. Rates for QSound functions #1, #2, #3 and #4 are 44100, 22050, 11025 and 8000Hz respectively. IMA ADPCM and FastSpeech codecs are formed of
.B two functions,
the first one is for capture and the second one is for playback. A\-law and u\-law codecs have only
.B one function.
To load for example function #2 from QSound microcode file (decoder for 22050Hz), run as follows:
.IP "" 4
% cspctl load \-f 2 \-d QSound wfm0001a.csp
.PP
After successful loading, /proc/asound/SB16/cspD2 will show:
.IP "" 4
.Sp
.nf
Creative Signal Processor [v1.0]
State: \-x\-\-L
Codec: QSound [func #2]
Sample rates: 22050Hz
QSound decoder disabled
.Sp
.fi
.PP
With QSound decoder microcode loaded, all simple PCM file formats can be played with QSound 180 degree positioning applied. QSound element is dynamically added into mixer structure as
.I 3DEffect1\-space
element. It will only show if support for CSP has been compiled into ALSA drivers, CSP chip has been found, and QSound codec is loaded into CSP. When enabled, QSound position can be dynamically changed by mixer slider, and proc interface should follow the state of CSP chip and its current QSound position:
.IP "" 4
.Sp
.nf
Creative Signal Processor [v1.0]
State: Qx\-RL
Codec: QSound [func #2]
Sample rates: 22050Hz
QSound decoder enabled
Processing 16bit mono PCM samples
Qsound position: left = 0x9, right = 0x9
.Sp
.fi
.PP
Driver supports autoloading of u\-Law, A\-Law and Ima\-ADPCM hardware codecs. Autoloading is active only when there is no microcode loaded to CSP, and there is no need to preload appropriate *.csp files. To unload manually loaded microcode, you should run
.IP "" 4
% cspctl unload
.PP
If hardware codec microcode has been manually loaded, then CSP will support only loaded PCM format and autoloading will be disabled. In such case, proc interface will show loaded codec properties:
.IP "" 4
.Sp
.nf
Creative Signal Processor [v1.0]
State: \-x\-\-L
Codec: IMA\-ADPCM [func #2]
Sample rates: All
PCM format ID: 0x400000 (16bit/\-) [mono/stereo] [playback/\-]
.Sp
.fi
.PP
.SH COMMANDS
.TP
.B load
Load microcode from filename to CSP.
.TP
.B unload
Unload microcode from CSP.
.SH OPTIONS
.TP
.B \-h, \-\-help
Display a short help.
.TP
.B \-c card
Select a card number. Option defaults to 0.
.TP
.B \-f function
Select a function from microcode file. Option defaults to 1.
.TP
.B \-d description
Optional microcode description string.

.SH AUTHOR
Uros Bizjak <uros@kss\-loka.si>.







boa/CHANGES
** Changes from 0.94.13 to 0.94.14
 * SUBSTANTIALLY UPDATE AUTOCONF SYSTEM
   - all of the autoconf stuff is now located in the top level
     directory, and creates the appropriate Makefiles for building
     the system and documentation.
 * add [optional, Linux-only] check for sendfile system call
 * add new state, IOSHUFFLE, which utilizes sendfile (or
   emulates it using the request's buffer, otherwise)
 * make the default socket size 32K.  The client_stream_size stays constant
   at 8K, and the new 'buffer' size is 4K
   Note also that the new code uses the "default" socket buffer size
   it obtains with the first accept()ed socket.
 * use setjmp, instead of sigsetjmp (we dont mess with the signal mask
   so why bother!  it's one less syscall)
 * explicitly set pending_requests to 0 in select.c when sigterm_flag has
   been set, and also when the server socket is *not* set despite checking
   for it.
 * make default behavior be to leave stderr alone, but tie it
   to cgilog otherwise
 * add initial vhost_root support
 * remove support for normalize_path
 * use --disable-sendfile to disable sendfile support
   sendfile support for files > 100K is default, now.
   It's significantly faster and easier than the alternative
 * add "Host" header support to CGI environment
 * umask ~0770 before exec
 * backport hash.c from 0.95 and use fnv1a hash (see CREDITS)
 * don't set stderr close-on-exec
 * make some minor for very useful optimizations in the read/write
   loop for CGIs
 * make some adjustments for Solaris and other platforms
 * add yyerror function definition to boa_lexer.l
 * tie stdout to the access_log, unless there is no access_log,
   in which case tie it to /dev/null
 * use sensible defaults for umask (077) and (027 for CGI)
 * add and document new parameters (CGILog and CGIumask)
 * vast improvements to the cgi-test.cgi program, by Jon Nelson
   and Landon Curt Noll.
 * next 3 items by Don Mahurin (patches modified somewhat):
 * pidfile patch
 * default mime_types patch
 * NCSA environment environment variables wrapped in
   #ifdef USE_NCSA_CGI_ENV
 * fix some escaping issues with the directory indexer
   (Ulf Harnhammar)
 * poll or select available at compile time
 * as of 0.94.14rc2 through 0.94.14rc3:
 * remove support for lex/yacc (backported from 0.95)
 * add conditional support for http/1.1
 * as of 0.94.14rc4:
 * replace many #defines with enumerations
 * update specfile (thanks to Supreet Sethi via SourceForge)
 * support optional config file name argument
 * tentatively support Ranges
 * as of 0.94.14rc5 through 0.94.14rc7:
 * dummy install target and some cleanups for *BSD
 * update index_dir.c to be *really* safe WRT html and http escaping*
   * Note to self - please make that code less ugly
 * basic access (Allow/Deny) control
 * fix check-for-broken-setuid check in boa.c -- root *is* a legal user
 * as of 0.94.14rc9:
 * fix reversed argument passing in add_mime_type
 * disable-gunzip support
 * capitalization changes, etc.. to configure.in
 * use slightly newer AC_DEFINE 3-argument style in configure.in
 * add new member of struct, bytes_written, and use it to more accurately
   report the number of bytes actually written to the socket.
 * When sendfile(2) reports ECONNRESET treat it like EPIPE, which is to
   say silently shut it down and don't be noisy about logging it.
 * wrap sa_family_t typedef in
   #ifdef DONT_HAVE_SA_FAMILY_T
   to deal with non-POSIX (1g) systems (Cygwin?)
 * add req->bytes_written member variable, and use it instead of filepos
 * apply Peter Korsgaard's "configure --help" patch
 * apply Peter Korsgaard's "configure --disable-gunzip" patch
 * use newer 3-argument AC_DEFINE
 * update config.sub and config.guess
 * clean up and update configure.in somewhat
 * don't close stderr in terminal signal handlers
 * free range pool and server_name in (final) sigterm handler
 * refactor select and poll loops
 * as of 0.94.14rc10:
 * change to using PF_ prefix instead of AF_ prefix (PF_ is POSIX?)
 * fix copying too-much-memory (read-side, write-side was OK) in
   ascii_sockaddr.  Also optimize slightly (note, may actually be a wee
   slower due to strlen check -- is there a way around this?)
 * fix a few shadow variable problems and improve select and poll loops
 * check for and use madvise (may or may not help)
 * update depends (the cause of SIGSEGs when ./configure was re-run to
   switch from select to poll, *without* running make clean)
 * always remove select.o and poll.o (and access.o) because these are
   the usual conditional files.
 * if we aren't using IPv6, define BOA_NI_MAXHOST to 20. 1025 is
   far too huge for a single IPv4 IP address in dotted-quad notation.
 * move access_init to earlier in the config reading
 * allow a MAX_FILE_MMAP value of 0 to mean "always mmap"
 * as of 0.94.14rc11:
 * fix 'HEAD' requests (bug introduced in rc6)
 * as of 0.94.14rc12:
 * *huge* patch to try to eliminate shadow variables, use unsigned ints
   instead of signed ints, and "const" where appropriate.
 * check for and use __func__, a C99 construct that is used
   in the DIE and WARN macros to also describe the name of the current
   function.
 * fix a very minor IPv6 issue, and include netdb.h in compat.h so that
   NI_MAXHOST is defined for IPv6
 * when we can't mmap a file, fall back to IOSHUFFLE.  If we couldn't
   mmap the file due to an error in mmap or madvise, report the error,
   otherwise it is safe to assume we simply ran out of hash table space.
   This was reported by Davide Pagnin
 * The following 2 changes borrowed from Hydra (which is itself
   based on Boa):
   * Some optimizations in HTTP header parsing.
   * Added DefaultCharset configuration directive. If given, the default
     character set will be appended to the Content-Type for all
     'text' mime types.
 * update boa.texi with some missing configuration directives
 * as of 0.94.14rc13:
 * force select.c, poll.c, and access.c to always be included in

alsa-tools/sb16_csp/cspctl.c
/*
 *  SB16/AWE32 Creative Signal Processor (ASP/CSP) control program
 *
 *  Copyright (c) 2000 by Uros Bizjak <uros@kss-loka.si>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 *
 */

#include <alsa/asoundlib.h>
#include <getopt.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <assert.h>
#include <string.h>
#include <errno.h>
#include <linux/ioctl.h>
#include <alsa/sound/sb16_csp.h>

/* --- commands --- */
enum {
     COMMAND_LOAD,
     COMMAND_UNLOAD
};

snd_sb_csp_microcode_t microcode;

int card = 0;

static void error(const char *fmt,...)
{
	va_list va;

	va_start(va, fmt);
	fprintf(stderr, "Error: ");
	vfprintf(stderr, fmt, va);
	fprintf(stderr, "\n");
	va_end(va);
}

static void help(char *command)
{
	fprintf(stderr,
		"Usage: %s command [-options] <filename>\n"
		"\nAvailable commands:\n"
		"  load    load microcode from filename\n"
		"  unload  unload microcode from CSP\n"
		"\nAvailable options:\n"
		"  -h,--help        this help\n"
		"  -c card          select card number, default = 0\n"
		"  -f function      select CSP function #, defaults to 1\n"
		"  -d description   optional codec description string\n"
		, command);
}

static int csp_command (int idx, int dev, int command, char *filename)
{
	int fd, err;
	snd_hwdep_t *handle;
	char name[16];

	sprintf(name, "hw:%i,%i", idx, dev);

	/* open CSP hwdep device */
	if ((err = snd_hwdep_open(&handle, name, O_WRONLY)) < 0) {
		error("CSP open (%i-%i): %s", idx, dev, snd_strerror(err));
		exit(1);
	}

	switch (command) {
	case COMMAND_LOAD:
		/* open microcode file */
		if ((fd = open(filename, O_RDONLY, 0)) == -1) {
			perror(filename);
			exit(1);
		}
		/* read microcode to buffer */
		if (read(fd, &microcode.data, sizeof(microcode.data)) < 0) {
			error("%s: read error", filename);
			exit(1);
		}
		close(fd);

		/* load microcode to CSP */
		if (snd_hwdep_ioctl(handle, SNDRV_SB_CSP_IOCTL_LOAD_CODE, &microcode) < 0) {
			error("unable to load microcode");
			exit(1);
		}
		break;
	case COMMAND_UNLOAD:
		/* unload microcode from CSP */
		if (snd_hwdep_ioctl(handle, SNDRV_SB_CSP_IOCTL_UNLOAD_CODE, NULL) < 0) {
			error("unable to unload microcode");
			exit(1);
		}
	}

	snd_hwdep_close(handle);
	return 0;
}

int main(int argc, char *argv[])
{
	int dev;
	int command, c;
	int err;

	char card_id[32];

	snd_ctl_t *ctl_handle;
	snd_ctl_card_info_t *card_info;
	snd_hwdep_info_t *hwdep_info;

	snd_ctl_card_info_alloca(&card_info);
	snd_hwdep_info_alloca(&hwdep_info);

	if (argc > 1 && !strcmp(argv[1], "--help")) {
		help(argv[0]);
		return 0;
	}

	strcpy (microcode.info.codec_name, "UNKNOWN");
	microcode.info.func_req = 1;
	while ((c = getopt(argc, argv, "hc:f:d:")) != EOF) {
		switch (c) {
		case 'h':
			help(argv[0]);
			return 0;
		case 'c':
			{
				card = snd_card_get_index(optarg);
				if (card < 0 || card > 31) {
					error ("wrong -c argument '%s'\n", optarg);
					return 1;
				}
			}
			break;
		case 'f':
			microcode.info.func_req = atoi(optarg);
			if ((microcode.info.func_req < 1) || (microcode.info.func_req > 4)) {
				error("value %i for function number is invalid",
				      microcode.info.func_req);
				return 1;
			}
			break;
		case 'd':
			if (strlen(optarg) > 16) {
				error("codec description '%s' too long", optarg);
				return 1;
			}
			strcpy(microcode.info.codec_name, optarg);
			break;
		default:
			return 1;
		}
	}
	if (optind >= argc) {
		error("please specify command");
		return 1;
	}
	if (!strcmp (argv[optind], "load")) {
	     command = COMMAND_LOAD;
	} else if (!strcmp (argv[optind], "unload")) {
	     command = COMMAND_UNLOAD;
	} else {
		error ("command should be either 'load' or 'unload'");
		return 1;
	}

	if ((command == COMMAND_LOAD) && (++optind >= argc)) {
		error ("missing microcode filename");
		return 1;
	}

	// Get control handle for selected card
	sprintf(card_id, "hw:%i", card);
	if ((err = snd_ctl_open(&ctl_handle, card_id, 0)) < 0) {
		error("control open (%s): %s", card_id, snd_strerror(err));
		return 1;
	}

	// Read control hardware info from card
	if ((err = snd_ctl_card_info(ctl_handle, card_info)) < 0) {
		error("control hardware info (%s): %s", card_id, snd_strerror(err));
		exit(1);
	}

	// CSP chip is present only on SB16 and SB AWE cards
	if (strcmp(snd_ctl_card_info_get_driver(card_info), "SB16") != 0 &&
	    strcmp(snd_ctl_card_info_get_driver(card_info), "SB AWE") != 0) {
		error("not a SB16 or SB AWE type card");
		exit(1);
	}

	// find CSP hardware dependant device and execute command
	dev = -1;
	err = 1;
	while (1) {
		if (snd_ctl_hwdep_next_device(ctl_handle, &dev) < 0)
			error("hwdep next device (%s): %s", card_id, snd_strerror(err));
		if (dev < 0)
			break;
		snd_hwdep_info_set_device(hwdep_info, dev);
		if (snd_ctl_hwdep_info(ctl_handle, hwdep_info) < 0) {
			if (err != -ENOENT)
				error("control hwdep info (%s): %s", card_id, snd_strerror(err));
			continue;
		}
		if (snd_hwdep_info_get_iface(hwdep_info) == SND_HWDEP_IFACE_SB16CSP) {
			err = csp_command (card, dev, command, argv[optind]);
			break;
		}
	}
	if (err)
		error("no CSP device present");

	snd_ctl_close(ctl_handle);
	return 0;
}
